agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,700 @@
+	switch keyType {
+	case "rsa":
+		switch keyBits {
+		case 2048:
+			return acmpca.KeyAlgorithmRsa2048, acmpca.SigningAlgorithmSha256withrsa, nil
+		case 4096:
+			return acmpca.KeyAlgorithmRsa4096, acmpca.SigningAlgorithmSha256withrsa, nil
+		default:
+			return "", "", fmt.Errorf("AWS PCA only supports RSA key lengths 2048"+
+				" and 4096, PrivateKeyBits of %d configured", keyBits)
+		}
+	case "ec":
+		if keyBits != 256 {
+			return "", "", fmt.Errorf("AWS PCA only supports P256 EC curve, keyBits of %d configured", keyBits)
+		}
+		return acmpca.KeyAlgorithmEcPrime256v1, acmpca.SigningAlgorithmSha256withecdsa, nil
+	default:
+		return "", "", fmt.Errorf("AWS PCA only supports P256 EC curve, or RSA"+
+			" 2048/4096. %s, %d configured", keyType, keyBits)
+	}
+}
+
+func (a *AWSProvider) createPCA() error {
+	pcaType := "ROOT" // For some reason there is no constant for this in the SDK
+	if !a.isPrimary {
+		pcaType = acmpca.CertificateAuthorityTypeSubordinate
+	}
+
+	keyAlg, signAlg, err := keyTypeToAlgos(a.config.PrivateKeyType, a.config.PrivateKeyBits)
+	if err != nil {
+		return err
+	}
+
+	uid, err := connect.CompactUID()
+	if err != nil {
+		return err
+	}
+	commonName := connect.CACN("aws", uid, a.clusterID, a.isPrimary)
+
+	createInput := acmpca.CreateCertificateAuthorityInput{
+		CertificateAuthorityType: aws.String(pcaType),
+		CertificateAuthorityConfiguration: &acmpca.CertificateAuthorityConfiguration{
+			Subject: &acmpca.ASN1Subject{
+				CommonName: aws.String(commonName),
+			},
+			KeyAlgorithm:     aws.String(keyAlg),
+			SigningAlgorithm: aws.String(signAlg),
+		},
+		RevocationConfiguration: &acmpca.RevocationConfiguration{
+			// TODO support CRL in future when we manage revocation in Connect more
+			// generally.
+			CrlConfiguration: &acmpca.CrlConfiguration{
+				Enabled: aws.Bool(false),
+			},
+		},
+		// uid is unique each PCA we create so use it as an idempotency string. We
+		// don't actually retry on failure yet but might as well!
+		IdempotencyToken: aws.String(uid),
+		Tags: []*acmpca.Tag{
+			{Key: aws.String("consul_cluster_id"), Value: aws.String(a.clusterID)},
+			{Key: aws.String("consul_datacenter"), Value: aws.String(a.datacenter)},
+		},
+	}
+
+	a.logger.Printf("[DEBUG] creating new PCA %s", commonName)
+	createOutput, err := a.client.CreateCertificateAuthority(&createInput)
+	if err != nil {
+		return err
+	}
+
+	// wait for PCA to be created
+	newARN := *createOutput.CertificateAuthorityArn
+	describeInput := acmpca.DescribeCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(newARN),
+	}
+	_, err = a.pollLoop("Private CA", 2*time.Minute, func() (bool, string, error) {
+		describeOutput, err := a.client.DescribeCertificateAuthority(&describeInput)
+		if err != nil {
+			if err.(awserr.Error).Code() != acmpca.ErrCodeRequestInProgressException {
+				a.logger.Printf("[ERR] connect.ca.aws: error describing PCA: %s", err.Error())
```

Comment:
I thought it was a small antipattern to both log an error and also return the same error, due to the returned error almost definitely getting logged again one level up.

Reply:
It is, for some reason I thought this case warranted it because the top level one might not surface but now I look at it I'm not sure why that would be the case.